lib/veri_policy.ml

```diff
@@ -0,0 +1,171 @@
+open Core_kernel.Std
+open Graph
+open Bap.Std
+open Bap_traces.Std
+
+module Action = struct
+  type t = string
+  let of_string = ident
+  let is act act' = String.compare act act' = 0
+end
+
+module Field = struct
+  type t = string
+  let of_string = ident
+  let empty = ""
+  let is_empty x = x = empty
+end
+
+type field = Field.t
+type action = Action.t
+type event = Trace.event
+type events = Value.Set.t
+type set_pair = events * events 
+type deny_error = set_pair
+type r = (set_pair, deny_error) Result.t
+
+type trial = {
+  regexp : Re.re;
+  field  : Field.t
+}
+
+type t = {
+  action : action;
+  insn   : trial;
+  left   : trial;
+  right  : trial;
+} [@@deriving fields]
+
+let skip = Action.of_string "SKIP"
+let deny = Action.of_string "DENY"
+let is_skip = Action.is skip
+let is_deny = Action.is deny
+
+let make_trial s = 
+  let field = match s with 
+    | None -> Field.empty
+    | Some s -> s in {
+    regexp = Re.compile (Re_posix.re field);
+    field;
+  }
+
+let create ?insn ?left ?right action : t = {
+  action; 
+  insn  = make_trial insn; 
+  left  = make_trial left; 
+  right = make_trial right; 
+}
+
+let sat e s = Re.execp e.regexp s
+let sat_event e ev = sat e (Value.pps () ev)
+let sat_events (e, ev) (e', ev') = sat_event e ev && sat_event e' ev'
+
+module G = struct
+  type g = trial * event array
+  type t = g * g
+  module V = struct
+    type t = Source | Sink | Person of int | Task of int [@@deriving compare]
+    let hash = Hashtbl.hash
+    let equal x y = compare x y = 0
+  end
+  module E = struct
+    type label = unit
+
+    type t = {src : V.t; dst : V.t} [@@deriving fields]
+    let make src dst = {src; dst}
+    let label _ = ()
+  end
+  type dir = Succ | Pred
+
+  let iter dir f ((expect, workers), (expect', jobs)) v = 
+    match v,dir with
+    | V.Source,Pred -> ()
+    | V.Source,Succ ->
+      Array.iteri workers ~f:(fun i _  ->
+          f @@ E.make V.Source (V.Person i))
+    | V.Sink,Pred ->
+      Array.iteri jobs ~f:(fun i _ ->
+          f @@ E.make (V.Person i) V.Sink)
+    | V.Sink,Succ -> ()
+    | V.Person i as p,Pred ->
+      f @@ E.make V.Source p
+    | V.Person i as p,Succ ->
+      Array.iteri jobs ~f:(fun j job ->
+          if sat_events (expect,workers.(i)) (expect', job) then 
```

Comment:
in our (unwritten) coding standard, we put `then` underneath the `if` unless the whole expression fits into the line.